Snowflake.Data/Core/FileTransfer/EncryptionProvider.cs

```diff
@@ -238,28 +257,36 @@ private static byte[] decryptFileKey(byte[] masterKey, byte[] unencryptedFileKey
         /// <param name="key">The encryption key.</param>
         /// <param name="iv">The encryption IV or null if it needs to be generated.</param>
         /// <returns>The decrypted bytes stream</returns>
-        private static Stream CreateDecryptedBytesStream(
+        private static StreamPair CreateDecryptedBytesStream(
             string inFile,
             byte[] key,
             byte[] iv,
             FileTransferConfiguration transferConfiguration)
         {
-            Aes aes = Aes.Create();
-            aes.Key = key;
-            aes.Mode = CipherMode.CBC;
-            aes.Padding = PaddingMode.PKCS7;
-            aes.IV = iv;
-            
-            var targetStream = new FileBackedOutputStream(transferConfiguration.MaxBytesInMemory, transferConfiguration.TempDir);
-            CryptoStream cryptoStream = new CryptoStream(targetStream, aes.CreateDecryptor(), CryptoStreamMode.Write);
-
-            using(Stream inStream = File.OpenRead(inFile))
+            using (Aes aes = Aes.Create())
             {
-                inStream.CopyTo(cryptoStream);
-            }
-            cryptoStream.FlushFinalBlock();
+                aes.Key = key;
+                aes.Mode = CipherMode.CBC;
+                aes.Padding = PaddingMode.PKCS7;
+                aes.IV = iv;
 
-            return targetStream;
+                var targetStream = new FileBackedOutputStream(transferConfiguration.MaxBytesInMemory, transferConfiguration.TempDir);
+                using (var decryptor = aes.CreateDecryptor())
+                {
+                    CryptoStream cryptoStream = new CryptoStream(targetStream, decryptor, CryptoStreamMode.Write);
+                    using (Stream inStream = File.OpenRead(inFile))
+                    {
+                        inStream.CopyTo(cryptoStream);
+                    }
+                    cryptoStream.FlushFinalBlock();
+
+                    return new StreamPair
+                    {
+                        MainStream = targetStream,
+                        HelperStream = cryptoStream // cryptoStream cannot be closed here because it would close target stream as well
```

Comment:
If cryptoStream would use any unmanged memory we would have a memory leak.
We would like to close cryptoStream, but it would close targetStream as well.
The solution is to return both streams and close both of them when target stream is processed.

Reply:
cryptostream closes underlying stream which must be returned open, we would need to force it with
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptostream.-ctor?view=net-6.0#system-security-cryptography-cryptostream-ctor(system-io-stream-system-security-cryptography-icryptotransform-system-security-cryptography-cryptostreammode-system-boolean)
but might not compile on .net 4.7.1 per "applies to" section of the above link